internal/backend/workloadproxy/reconciler.go

```diff
@@ -6,225 +6,374 @@
-	registry.logger.Log(registry.logLevel, "remove LB", zap.String("cluster", cluster), zap.String("alias", alias))
+// Describe implements the [prometheus.Collector] interface.
+func (rec *Reconciler) Describe(ch chan<- *prometheus.Desc) { prometheus.DescribeByCollect(rec, ch) }
 
-	aliasToLB := registry.clusterToAliasToLBStatus[cluster]
-	lbSts := aliasToLB[alias]
+// Collect implements the [prometheus.Collector] interface.
+func (rec *Reconciler) Collect(mcs chan<- prometheus.Metric) { rec.idleConnections.Collect(mcs) }
 
-	if lbSts != nil {
-		if err := lbSts.Close(); err != nil {
-			registry.logger.Error("failed to close LB", zap.String("cluster", cluster), zap.String("alias", alias), zap.Error(err))
+func asRemotes(hosts []string, p port, l *zap.Logger) iter.Seq[*remote] {
+	return func(yield func(*remote) bool) {
+		for _, h := range hosts {
+			if !yield(&remote{Logger: l, Addr: h, AddrPort: net.JoinHostPort(h, string(p))}) {
+				return
+			}
 		}
 	}
+}
 
-	delete(aliasToLB, alias)
-	delete(registry.aliasToCluster, alias)
+// setEvictionFunc sets the eviction function for the activeProbes cache. It will shut down the active load balancer
+// probe.
+func (rec *Reconciler) setEvictionFunc() func() {
+	l := rec.logger
+	lvl := rec.logLevel
+	fn := func(_ context.Context, _ ttlcache.EvictionReason, i *ttlcache.Item[resource.ID, remoteList]) {
+		l.Log(lvl, "shutting down active probe LB", zap.String("cluster", i.Key()))
 
-	if len(aliasToLB) == 0 {
-		delete(registry.clusterToAliasToLBStatus, cluster)
+		i.Value().Shutdown()
 	}
+
+	return rec.activeProbes.OnEviction(fn)
 }
 
-// hostPortForAlias returns a unique IP:port for the given cluster and alias.
-//
-// The value is arbitrary, as it uses in-memory transport, never reached via the network.
-func (registry *Reconciler) hostPortForAlias(clusterID resource.ID, alias string) string {
-	return fmt.Sprintf("%s_%s:4242", clusterID, alias)
+// Shutdown stops the Reconciler and all the active probes.
+func (rec *Reconciler) Shutdown() {
+	rec.shutdown()
+}
+
+type remoteList = *upstream.List[*remote]
+
+// Alias is an alias for the cluster specific port.
+type alias string
+
+// String implements the Stringer interface.
+func (a alias) String() string { return string(a) }
+
+type port string
+
+func (p port) String() string { return string(p) }
+
+type roundTripperFunc func(request *http.Request) (*http.Response, error)
+
+func (r roundTripperFunc) RoundTrip(request *http.Request) (*http.Response, error) { return r(request) }
+
+func wrapWithMetrics(next *http.Transport, rec *Reconciler) roundTripperFunc {
+	return func(r *http.Request) (*http.Response, error) {
+		trace := &httptrace.ClientTrace{
+			GotConn: func(info httptrace.GotConnInfo) {
+				if info.WasIdle {
+					rec.idleConnections.Dec()
+				}
+
+				rec.inUseConnections.Inc()
+			},
+			PutIdleConn: func(err error) {
+				if err == nil {
+					rec.idleConnections.Inc()
```

Comment:
I'm curious if this is accurate in case idle connection is closed on timeout - there will be no httptrace here?

Reply:
Yeah, I agree that looks like a problem. I tried to fix this in the new version - instead of idleConnections we have total connections and finalizer attached to them. The difference between `inUse` and `total` should tell us how many are idle (more or less).